storm-client/src/jvm/org/apache/storm/topology/PersistentWindowedBoltExecutor.java

```diff
@@ -0,0 +1,563 @@
+    private static final Logger LOG = LoggerFactory.getLogger(PersistentWindowedBoltExecutor.class);
+    private final IStatefulWindowedBolt<T> statefulWindowedBolt;
+    private transient TopologyContext topologyContext;
+    private transient OutputCollector outputCollector;
+    private transient WindowState<Tuple> state;
+    private transient boolean stateInitialized;
+    private transient boolean prePrepared;
+
+    public PersistentWindowedBoltExecutor(IStatefulWindowedBolt<T> bolt) {
+        super(bolt);
+        statefulWindowedBolt = bolt;
+    }
+
+    @Override
+    public void prepare(Map<String, Object> topoConf, TopologyContext context, OutputCollector collector) {
+        List<String> registrations = (List<String>) topoConf.getOrDefault(Config.TOPOLOGY_STATE_KRYO_REGISTER, new ArrayList<>());
+        registrations.add(ConcurrentLinkedQueue.class.getName());
+        registrations.add(LinkedList.class.getName());
+        registrations.add(AtomicInteger.class.getName());
+        registrations.add(EventImpl.class.getName());
+        registrations.add(WindowPartition.class.getName());
+        registrations.add(DefaultEvictionContext.class.getName());
+        topoConf.put(Config.TOPOLOGY_STATE_KRYO_REGISTER, registrations);
+        prepare(topoConf, context, collector,
+            getWindowState(topoConf, context),
+            getPartitionState(topoConf, context),
+            getWindowSystemState(topoConf, context));
+    }
+
+    // package access for unit tests
+    void prepare(Map<String, Object> topoConf, TopologyContext context, OutputCollector collector,
+                 KeyValueState<Long, WindowPartition<Tuple>> windowState,
+                 KeyValueState<String, Deque<Long>> partitionState,
+                 KeyValueState<String, Optional<?>> windowSystemState) {
+        init(topoConf, context, collector, windowState, partitionState, windowSystemState);
+        doPrepare(topoConf, context, new NoAckOutputCollector(collector), state, true);
+        Map<String, Optional<?>> wstate = new HashMap<>();
+        windowSystemState.forEach(s -> wstate.put(s.getKey(), s.getValue()));
+        restoreState(wstate);
+    }
+
+    @Override
+    protected void start() {
+        if (stateInitialized) {
+            super.start();
+        } else {
+            LOG.debug("Will invoke start after state is initialized");
+        }
+    }
+
+    @Override
+    public void execute(Tuple input) {
+        if (!stateInitialized) {
+            throw new IllegalStateException("execute invoked before initState with input tuple " + input);
+        }
+        super.execute(input);
+        // StatefulBoltExecutor does the actual ack when the state is saved.
+        outputCollector.ack(input);
+    }
+
+    @Override
+    public void initState(T state) {
+        if (stateInitialized) {
+            LOG.warn("State is already initialized. Ignoring initState");
```

Comment:
If a worker crashes, all worker's state are rolled back and an initState message is sent across the topology so that crashed workers can initialize their state. This is already handled in `StatefulBoltExecutor` that wraps the `PersistentWindowedBoltExecutor` so here its Ok to throw an exception. Will address this and revisit the others.